File: app/src/main/java/uk/co/alt236/apkcompare/app/output/html/FileFactory.java
package uk.co.alt236.apkcompare.app.output.html;

import uk.co.alt236.apkcompare.app.comparators.results.ComparisonResult;

import java.io.File;
import java.util.Locale;

class FileFactory {
    private static final String FOLDER_SUFFX = "_files";
    private static final String FILE_PROTOCOL = "file:/";

    File getResultHtmlFile(final ComparisonResult result,
                           final File rootFile) {
        final String newFilename = result.getTitle().replaceAll("[^\\w.-]", "_").toLowerCase(Locale.US);
        return new File(rootFile + FOLDER_SUFFX, newFilename + ".html");
    }


    String getRelativeLink(final ComparisonResult result,
                           final File rootFile) {

        final File resultFile = getResultHtmlFile(result, new File(rootFile.getName()));
        return resultFile.toString().replace("./", "");
    }

}

File: app/src/main/java/uk/co/alt236/apkcompare/app/output/html/FileFactory.kt
package uk.co.alt236.apkcompare.app.output.html

import uk.co.alt236.apkcompare.app.comparators.results.ComparisonResult
import java.io.File
import java.util.*

internal class FileFactory(val rootFile: File) {
    private val rootFileName: String = rootFile.name

    fun getResultHtmlFile(result: ComparisonResult): File {
        return getResultHtmlFile(result, rootFile)
    }

    fun getLinkRelativeToRoot(result: ComparisonResult): String {
        val resultFile = getResultHtmlFile(result, File(rootFileName))
        return resultFile.toString().replace("./", "")
    }

    fun getLinkRelativeToFiles(result: ComparisonResult): String {
        return "${result.title.sanitise()}.html"
    }

    private fun getResultHtmlFile(result: ComparisonResult, baseFile: File): File {
        val newFilename = result.title.sanitise()
        return File("$baseFile$FOLDER_SUFFIX", "$newFilename.html")
    }


    private fun String.sanitise(): String {
        return this.replace(saneRegex, "_")
                .toLowerCase(Locale.US)
    }

    private companion object {
        private val saneRegex = "[^\\w.-]".toRegex()
        private const val FOLDER_SUFFIX = "_files"
    }
}
